### PR TITLE
apimachinery: fix Unstructured deepcopy panic

### PR DESCRIPTION
**Pull request by sttts**
Fixes https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/54484/pull-kubernetes-unit/61498/#k8siokubernetescmdkube-apiserverapptesting-testcrd

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go [hunk 1]
```diff
@@ -490,6 +490,10 @@ func (u *Unstructured) GetDeletionGracePeriodSeconds() *int64 {
 }
 
 func (u *Unstructured) SetDeletionGracePeriodSeconds(deletionGracePeriodSeconds *int64) {
+	if deletionGracePeriodSeconds == nil {
+		u.setNestedField(nil, "metadata", "deletionGracePeriodSeconds")
```
Comment (ash2k) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go]: My PR (https://github.com/kubernetes/kubernetes/pull/51940) removes the value if it's `nil` which I think is the right thing to do in this particular case.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_test.go [hunk 2]
```diff
@@ -75,4 +75,34 @@ func TestNilDeletionTimestamp(t *testing.T) {
...
+	if del != nil {
+		t.Errorf("unexpected non-nil deletion grace period: %v", del)
+	}
+	u.SetDeletionGracePeriodSeconds(u.GetDeletionGracePeriodSeconds())
+	del = u.GetDeletionGracePeriodSeconds()
+	if del != nil {
+		t.Errorf("unexpected non-nil deletion grace period: %v", del)
+	}
+	metadata := u.Object["metadata"].(map[string]interface{})
+	deleteGracePeriod := metadata["deletionGracePeriod"]
```
Comment (ash2k) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_test.go]: These two tests should probably distinguish between `present, nil` and `no value in map` cases and expect something specific.

Comment (ash2k) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_test.go]: Both are `+optional` and `omitempty` so there should be no value in the map.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go (hunk 1 above)
Comment (sttts) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go]: Yes, this is the quick fix only for the panic.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_test.go (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_test.go]: Am not against this, as a follow-up preferibly.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go (hunk 1 above)
Comment (deads2k) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go]: Ah, we have been round this a few times.  Found it again here: https://github.com/kubernetes/kubernetes/pull/53464/files#r148250781 .  Is really any harder to remove the entry?

Comment (deads2k) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go]: Just to go on record.  I like removing the key entirely from omitempty fields so serializations are closer to a match.

Comment (sttts) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go]: @deads2k the PR #51940 by @ash2k does exactly this. Great cleanup.

**Comment by sttts** [comment]
Superseeded by https://github.com/kubernetes/kubernetes/pull/51940.

**Pull request closed without merging**